Add JWT cache to JWT realm.

justincr-elastic opened the pull request:
This simple JWT cache mirrors the design from `PkiRealm` methods `PkiRealm()`, `authenticate()`, `expire()`, and `expireAll()`.

JWT cache settings are prefixed by `jwt.`.
```
jwt.cache.ttl
jwt.cache.max_users
```

For example, a planned client PKI cache in future will use a different prefix, possibly `client.`.
```
client.cache.ttl
client.cache.max_users
```

justincr-elastic committed:
Add JWT cache to JWT realm.

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

elasticsearchmachine commented:
Hi @justincr-elastic, I've created a changelog YAML for you.

justincr-elastic committed:
Update docs/changelog/84842.yaml

justincr-elastic committed:
Remove jwt.cache.hash_algo. Use ByteKey with fixed SHA256 for cache keys.

justincr-elastic committed:
Merge remote-tracking branch 'fork/feature/jwt-realm-jwt-cache' into feature/jwt-realm-jwt-cache

justincr-elastic committed:
Revert tests.

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-jwt-cache

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:
```diff
@@ -310,22 +368,29 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
+                final User user = this.jwtCache.get(jwtCacheKey);
+                if (user == null) {
+                    LOGGER.trace("Realm [" + super.name() + "] JWT cache miss token=[" + tokenPrincipal + "] key=[" + jwtCacheKey + "].");
+                } else {
```
This will allow authentication using expired JWTs

justincr-elastic replied:
Wrapped User in an ExpiringUser record. Store User with exp+allowedSkew. Check exp against wall clock during cache hit.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java:
```diff
@@ -229,6 +227,20 @@ public static ClientAuthenticationType parse(String value, String settingKey) {
+
+    public static final Setting.AffixSetting<Integer> JWT_CACHE_MAX_USERS = Setting.affixKeySetting(
+        RealmSettings.realmSettingPrefix(TYPE),
+        "jwt.cache.max_users",
```
This setting controls the number of entries in the cache, which may not be the number of users.
I think it should be
`jwt.cache.size` (like the DLS bitset and automata caches)
or `jwt.cache.max_keys` (like the api-key cache)

justincr-elastic replied:
Renamed to `jwt.cache.size`.

justincr-elastic committed:
Merge branch 'master' of file:///c/GitHub/mirrors/elasticsearch into feature/jwt-realm-jwt-cache

justincr-elastic committed:
Merge remote-tracking branch 'fork/feature/jwt-realm-jwt-cache' into feature/jwt-realm-jwt-cache

justincr-elastic committed:
Rename jwt.cache.max_users to jwt.cache.size.

justincr-elastic committed:
Use CacheIteratorHelper new method removeValuesIf().

justincr-elastic committed:
Checkstyle.

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-jwt-cache

justincr-elastic committed:
Handle expired JWTs in cache.

justincr-elastic commented:
I think I addressed all of the review comments. Setting is renamed, I am using CacheIteratorHelper, and cache values are wrapped in an ExpiringUser(User,expDate) record. Insert into cache sets expDate = Date[ JWT(expMillis) + Realm(skewMillis) ]. Cache hits are checked against wall clock, to handle JWTs that expire sooner than cache expiration setting.

justincr-elastic committed:
Merge branch 'master' of file:///c/GitHub/mirrors/elasticsearch into feature/jwt-realm-jwt-cache

justincr-elastic commented:
@elasticmachine update branch

Unrelated test failure. Update from master and rebuild.

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-jwt-cache

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-jwt-cache

justincr-elastic commented on docs/changelog/84842.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 84842
+summary: Add JWT cache to JWT realm
+area: Authentication
+type: feature
```
Changed label. Resolved.

justincr-elastic committed:
Delete docs/changelog/84842.yaml

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:
```diff
@@ -272,11 +306,19 @@ public void lookupUser(final String username, final ActionListener<User> listene
     public void expire(final String username) {
         this.ensureInitialized();
+        LOGGER.trace("Expiring JWT cache entries for realm [" + super.name() + "] principal=[" + username + "]");
+        this.jwtCacheHelper.removeValuesIf(expiringUser -> expiringUser.user.principal().equals(username));
```
Ok. I think it makes sense to add a null check in `CacheIteratorHelper`. Since `removeValuesIf` is a copy of `removeKeysIf`, that implies both need a null check.

https://github.com/elastic/elasticsearch/blob/45f2851562e68c393954df5be942e1764e8dbe09/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/CacheIteratorHelper.java#L48-L59

https://github.com/elastic/elasticsearch/blob/45f2851562e68c393954df5be942e1764e8dbe09/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/CacheIteratorHelper.java#L61-L72

justincr-elastic committed:
Add null checks for jwtCacheHelper.

justincr-elastic committed:
Merge remote-tracking branch 'fork/feature/jwt-realm-jwt-cache' into feature/jwt-realm-jwt-cache
